pkgs/tools/package-management/nix/common-meson.nix

```
@@ -0,0 +1,284 @@
+  aws-sdk-cpp,
+  withLibseccomp ? lib.meta.availableOn stdenv.hostPlatform libseccomp,
+  libseccomp,
+
+  confDir,
+  stateDir,
+  storeDir,
+
+  # passthru tests
+  pkgsi686Linux,
+  pkgsStatic,
+  runCommand,
+  pkgs,
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "nix";
+
+  version = "${version}${suffix}";
+  VERSION_SUFFIX = suffix;
+
+  inherit src patches;
+
+  outputs =
+    [
+      "out"
+      "dev"
+    ]
+    ++ lib.optionals enableDocumentation [
+      "man"
+      "doc"
+    ];
+
+  hardeningEnable = lib.optionals (!stdenv.hostPlatform.isDarwin) [ "pie" ];
+
+  hardeningDisable = [
+    "shadowstack"
+  ] ++ lib.optional stdenv.hostPlatform.isMusl "fortify";
+
+  nativeCheckInputs =
+    [
+      git
+      man
+    ]
+    ++ lib.optionals (!stdenv.isDarwin) [
+      perlPackages.Test2Harness
+    ];
+
+  nativeBuildInputs =
+    [
+      bison
+      cmake
+      flex
+      jq
+      meson
+      ninja
+      pkg-config
+      rsync
+    ]
+    ++ lib.optionals enableDocumentation [
+      (lib.getBin lowdown-unsandboxed)
+      mdbook
+      mdbook-linkcheck
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isLinux [
+      util-linuxMinimal
+    ];
+
+  buildInputs =
+    [
+      boost
+      brotli
+      bzip2
+      curl
+      editline
+      gtest
+      libarchive
+      libgit2
+      libsodium
+      lowdown
+      openssl
+      rapidcheck
+      sqlite
+      toml11
+      xz
+      perl
+    ]
+    ++ lib.optionals enableDocumentation [
+      python3
+      doxygen
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [
+      Security
+    ]
+    ++ lib.optionals stdenv.hostPlatform.isx86_64 [
+      libcpuid
+    ]
+    ++ lib.optionals withLibseccomp [
+      libseccomp
+    ]
+    ++ lib.optionals withAWS [
+      aws-sdk-cpp
+    ]
+    ++ lib.optional (stdenv.hostPlatform.isDarwin) [
+      darwin.apple_sdk.libs.sandbox
+    ];
+
+  propagatedBuildInputs = [
+    boehmgc
+    nlohmann_json
+  ];
+
+  postPatch = ''
+    patchShebangs --build tests
+  '';
+
+  preConfigure =
+    # Copy libboost_context so we don't get all of Boost in our closure.
+    # https://github.com/NixOS/nixpkgs/issues/45462
+    lib.optionalString (!enableStatic) ''
+      mkdir -p $out/lib
+      cp -pd ${boost}/lib/{libboost_context*,libboost_thread*,libboost_system*} $out/lib
+      rm -f $out/lib/*.a
+      ${lib.optionalString stdenv.hostPlatform.isLinux ''
+        chmod u+w $out/lib/*.so.*
+        patchelf --set-rpath $out/lib:${lib.getLib stdenv.cc.cc}/lib $out/lib/libboost_thread.so.*
+      ''}
+    '';
+
+  dontUseCmakeConfigure = true;
+
+  mesonFlags =
+    [
+      (lib.mesonBool "bindings" false)
+      (lib.mesonOption "perl:dbi_path" "${perlPackages.DBI}/${perl.libPrefix}")
+      (lib.mesonOption "perl:dbd_sqlite_path" "${perlPackages.DBDSQLite}/${perl.libPrefix}")
+      (lib.mesonEnable "perl:tests" (finalAttrs.doCheck && !stdenv.hostPlatform.isDarwin))
```

I think, because of complicated stdenv reasons, this is supposed to also use `finalPackage`. e.g. https://github.com/NixOS/nixpkgs/pull/271241#issuecomment-1834596148

```suggestion
      (lib.mesonEnable "perl:tests" (finalAttrs.finalPackage.doCheck && !stdenv.hostPlatform.isDarwin))
```